### PR TITLE
vtgate: record warning for partially successful cross-shard commits

### PR DESCRIPTION
**Pull request by maxenglander**
## Description

When using `multi` transaction mode (the default), it is possible for Vitess to successfully commit to one shard, but fail to commit to a subsequent shard, thus breaking the atomicity of a multi-shard transaction.

This PR updates VTGate to report partial-success commits in warnings, e.g.:

```mysql
mysql> commit;
ERROR 1317 (70100): target: customer.-80.primary: vttablet: rpc error: code = Aborted desc = transaction 1703182545849001001: ended at 2023-12-21 14:07:41.515 EST (exceeded timeout: 30s) (CallerID: userData1)
mysql> show warnings;
+---------+------+-----------------------------------------------------+
| Level   | Code | Message                                             |
+---------+------+-----------------------------------------------------+
| Warning |  301 | multi-db commit failed after committing to 1 shards |
+---------+------+-----------------------------------------------------+
1 row in set, 1 warning (0.00 sec)
```

Also increments the `vtgate_warnings{type="NonAtomicCommit"}` count.

## Related Issue(s)

This PR is intended as a partial, approximate solution for https://github.com/vitessio/vitess/issues/14815.

## Checklist

Backport labels not needed.

-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added https://github.com/vitessio/website/pull/1655

### CODE REVIEW

**Review by systay — APPROVED** [review]

**Comment by harshit-gangal** [comment]
The idea behind raising a warning is good. Should we also provide the shards where the commit was successful? So, the users can do manual fixes if needed.

**Review by harshit-gangal — COMMENTED** [review]
LGTM, left one comment.

**Review by harshit-gangal — APPROVED** [review]

**Comment by maxenglander** [comment]
@harshit-gangal I'm happy to make that change! What's a reasonable upper bound on the number of shards to report in the warning? 256?

**Comment by harshit-gangal** [comment]
> @harshit-gangal I'm happy to make that change! What's a reasonable upper bound on the number of shards to report in the warning? 256?

256 would be a very large number to report in shards.
We can keep it to 8 or 16.

**Comment by harshit-gangal** [comment]
You should add this change in release notes here: https://github.com/vitessio/vitess/blob/main/changelog/19.0/19.0.0/summary.md

**Comment by maxenglander** [comment]
@harshit-gangal added a change log note and limited shards reported in warn to 16.

File: go/vt/vtgate/tx_conn.go [hunk 1]
```diff
@@ -132,8 +138,27 @@ func (txc *TxConn) commitNormal(ctx context.Context, session *SafeSession) error
...
+					nShards = nonAtomicCommitWarnMaxShards
+					elipsis = true
+				}
+				sNames := make([]string, nShards, nShards+1 /*...*/)
+				for j := 0; j < nShards; j++ {
+					sNames[j] = session.ShardSessions[j].Target.Shard
+				}
+				if elipsis {
+					sNames = append(sNames, "...")
+				}
```
Comment (harshit-gangal) [inline go/vt/vtgate/tx_conn.go:156]: can you add a test where this is covered?

**Pull request merged**